### PR TITLE
fix(events-targets): `ApiGateway` events target should accept IRestApi

### PR DESCRIPTION
**Pull request by n-k1**
### Issue

Closes [#16423](https://github.com/aws/aws-cdk/issues/16423).

### Reason for this change

The CDK construct to set an APIGatway as the target for an EventBridge rule only accepted RestApi. It should instead accept the interface IRestApi.

### Description of changes

This change was attempted once earlier, but [the PR](https://github.com/aws/aws-cdk/pull/16542) was closed.

Changed the `ApiGateway` target's constructor's first argument from `api.RestApi` to `api.IRestApi`.

To fix compatibility failures caused by the previous `restApi` being a public member of the class, I've followed the approach suggested [here](https://github.com/aws/aws-cdk/pull/16542#discussion_r713676896). Some other alternatives are suggested [here](https://github.com/aws/aws-cdk/pull/16542#issuecomment-925051255).

### Description of how you validated changes

Added a unit test that creates a dummy `SpecRestApi`. Also added an integ test that creates a `SpecRestApi` from a dummy OpenAPI spec and creates a rule on the default event bus with the APIGateway as target. Nothing is executed in the integ test. The assertion only checks if the target is created using the `ListTargetsByRule` API.

### Checklist
- [x] My code adheres to the [CONTRIBUTING GUIDE](https://github.com/aws/aws-cdk/blob/main/CONTRIBUTING.md) and [DESIGN GUIDELINES](https://github.com/aws/aws-cdk/blob/main/docs/DESIGN_GUIDELINES.md)

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

File: packages/aws-cdk-lib/aws-events-targets/lib/api-gateway.ts [hunk 1]
```diff
@@ -73,8 +73,31 @@ export interface ApiGatewayProps extends TargetBaseProps {
  * Use an API Gateway REST APIs as a target for Amazon EventBridge rules.
  */
 export class ApiGateway implements events.IRuleTarget {
+  private readonly _restApi: api.IRestApi;
 
-  constructor(public readonly restApi: api.RestApi, private readonly props?: ApiGatewayProps) {
+  /**
+   * @param restApi - IRestApi implementation to use as event target
+   * @param props - Properties to configure the APIGatway target
```
Comment (msambol) [inline packages/aws-cdk-lib/aws-events-targets/lib/api-gateway.ts:80]: nit: `ApiGateway`

File: packages/@aws-cdk-testing/framework-integ/test/aws-events-targets/test/api-gateway/integ.api-gateway.ts [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+// assert the results
+ruleDescription.expect(ExpectedResult.objectLike({
+  Targets: [
+    {
+      HttpParameters: {},
+    },
+  ],
+}));
+
+app.synth();
```
Comment (msambol) [inline packages/@aws-cdk-testing/framework-integ/test/aws-events-targets/test/api-gateway/integ.api-gateway.ts:84]: nit: newline

File: packages/@aws-cdk-testing/framework-integ/test/aws-events-targets/test/api-gateway/integ.api-gateway.ts [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+import * as events from 'aws-cdk-lib/aws-events';
+import * as api from 'aws-cdk-lib/aws-apigateway';
+import * as cdk from 'aws-cdk-lib';
+import { IntegTest, ExpectedResult } from '@aws-cdk/integ-tests-alpha';
+
+/*
+ * Stack verification steps :
+ * * aws stepfunctions start-execution --state-machine-arn <deployed state machine arn> : should return execution arn
+ * * aws stepfunctions describe-execution --execution-arn <execution-arn generated before> : should return status as SUCCEEDED
+ */
```
Comment (msambol) [inline packages/@aws-cdk-testing/framework-integ/test/aws-events-targets/test/api-gateway/integ.api-gateway.ts:11]: these aren't needed?

**Review by msambol — APPROVED** [review]
This LGTM. Moving to maintainer review. I'd like @rix0rrr to review this since he had feedback on the previous PR.

File: packages/aws-cdk-lib/aws-events-targets/lib/api-gateway.ts [hunk 4]
```diff
@@ -73,8 +73,31 @@ export interface ApiGatewayProps extends TargetBaseProps {
...
+  /**
+   * @param restApi - IRestApi implementation to use as event target
+   * @param props - Properties to configure the APIGateway target
+   */
+  constructor(restApi: api.IRestApi, private readonly props?: ApiGatewayProps) {
+    this._restApi = restApi;
+  }
+
+  /**
+   * @deprecated Use iRestApi instead
```
Comment (gracelu0) [inline packages/aws-cdk-lib/aws-events-targets/lib/api-gateway.ts:87]:
```suggestion
   * @deprecated Use IRestApi instead
```

File: packages/aws-cdk-lib/aws-events-targets/lib/api-gateway.ts [hunk 5]
```diff
@@ -73,8 +73,31 @@ export interface ApiGatewayProps extends TargetBaseProps {
...
+   */
+  public get restApi(): api.RestApi {
+    if (!(this._restApi instanceof api.RestApi)) {
```
Comment (gracelu0) [inline packages/aws-cdk-lib/aws-events-targets/lib/api-gateway.ts:90]: Can we add a test for this please?

**Review by gracelu0 — CHANGES_REQUESTED** [review]

File: packages/aws-cdk-lib/aws-events-targets/lib/api-gateway.ts (hunk 4 above)
Comment (n-k1) [reply packages/aws-cdk-lib/aws-events-targets/lib/api-gateway.ts:87]: the new getter is caller `iRestApi`

```
  public get iRestApi(): api.IRestApi {
```

Comment (gracelu0) [reply packages/aws-cdk-lib/aws-events-targets/lib/api-gateway.ts:87]:
```suggestion
   * @deprecated Use `iRestApi()` instead
```

**Review by gracelu0 — APPROVED** [review]
Thank you for your contribution!

**Pull request merged**